File: Merge-Sort/js/simple.js
const arr = []

function initData(count){
    for(let i = 0 ; i < count ; i++){
        arr.push(Math.floor(Math.random()*50-20))
    }
}

initData(6)

console.log(arr)

function mergeSort(arr){
    if(arr.length < 2) return arr

    const mid = Math.floor(arr.length/2),
        left = arr.slice(0,mid),
        right = arr.slice(mid);

    return merge(mergeSort(left),mergeSort(right))
}


function merge(left,right){
    const res = []
    let i = j = 0 ;

    while(i < left.length && j < right.length){
        if(left[i] <= right[j]){
            res.push(left[i])
            i++
        }else{
            res.push(right[j])
            j++
        }
    }

    while(i < left.length){
        res.push(left[i])
        i++
    }

    while(j < right.length){
        res.push(right[j])
        j++
    }

    return res
}

function merge_bak(left,right){
    const res = []

    while(left.length && right.length){
        if(left[0] <= right[0]){
            res.push(left.shift())
        }else{
            res.push(right.shift())
        }
    }

    while(left.length){
        res.push(left.shift())
    }

    while(right.length){
        res.push(right.shift())
    }

    return res
}


console.log(mergeSort(arr))

File: Merge-Sort/python/simple.py
import random

arr = [random.randint(-20,30) for _ in range(6)]

print(arr)

def mergeSort(arr):
    if len(arr) < 2: 
        return arr
    
    mid = int(len(arr)/2)
    left = arr[0:mid]
    right = arr[mid:]

    return merge(mergeSort(left),mergeSort(right))


def merge(left,right):
    res = []

    while left and right:
        if left[0] <= right[0]:
            res.append(left.pop(0))
        else:
            res.append(right.pop(0))

    while left:
        res.append(left.pop(0))

    while right:
        res.append(right.pop(0))


    return res


print(mergeSort(arr))